### DISCUSSION TITLE
Partial Generics for Collection Literals

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#8045] [Language Ideas] [answered]
Today I have to write code like this:
```
                    List<(int, string, bool, string)> Table = [
                        (1, "aoeu", true, "aoeu"),
                        (1, "aoeu", true, "aoeu"),
                        (1, "aoeu", true, "aoeu"),
                    ];
```

This sucks because if I want to add or reorder columns in my table, I also have to do that in the List<(???)> definition.

It would be so much nicer if I could do this:
```
                    List<> Table = [
                        (1, "aoeu", true, "aoeu"),
                        (1, "aoeu", true, "aoeu"),
                        (1, "aoeu", true, "aoeu"),
                    ];
```

**Comment by HaloFour** [comment]
Would this shake out of natural types for collection literals?

**Reply by CyrusNajmabadi** [reply]
yes.  but it wouldn't work if you did `ImmutableArray<> table = [ ... ]`.  We'd need a new system for that.  Fortunately, that is something being looked into.

**Reply by HaloFour** [reply]
Part of supporting natural types would be to support extension methods off of the collection literal, too, right?  In that you could write an extension method to support:

```cs
var array = [ ... ].ToImmutableArray();
```

For partial type inference: https://github.com/dotnet/csharplang/issues/1349

**Reply by KennethHoff** [reply]
Are extension method being looked at in conjunction with natural type? I realize that they are both being looked at, but they feel somewhat unrelated.

**Reply by CyrusNajmabadi** [reply, marked as answer]
Both are in the bucket of things we're looking at in this space :)
